### PR TITLE
feat(android): add auth0-android-major-migration skill

### PR DESCRIPTION
**Pull request by sanchitmehtagit**
### Description

Adds the **auth0-android-major-migration** skill — the Android counterpart to `auth0-swift-major-migration` (#104) — for upgrading an existing `com.auth0.android:auth0` (v3) integration to the next major version (v4).

The skill is **source-driven and grep-gated**: it fetches the v4 SDK's actual Kotlin source to confirm signatures, audits which Auth0 APIs the project actually calls, and applies only the breaking changes that affect real call sites — then builds until green and emits a migration summary.

**Argument-based version selection (improvement over the Swift sibling):**
- Pass a target version (e.g. `/auth0-android-major-migration 4.0.0`) — it's validated (published tag, within the next major, not a downgrade) and the skill **stops and asks** on failure.
- Omit it — the skill auto-resolves the latest v4 release (including pre-releases; resolves `4.0.0-beta.1` today).

**Prerequisite gate (RAPID "block until satisfied"):** minSdk 26, Java 17, Gradle 8.11.1+, AGP 8.10.1+, Kotlin 2.0.21 — checked and applied before any API migration.

**Breaking changes covered (all grep-gated):**

| § | Change |
|---|---|
| 7.1 | `PasskeyAuthProvider` removed → `AuthenticationAPIClient` passkey APIs |
| 7.2 | Management API removed (`UsersAPIClient`) → `// TODO:` + backend follow-up |
| 7.3 | Deprecated MFA methods removed → `mfaClient(mfaToken)` / `MfaApiClient` |
| 7.4 | `WebAuthProvider.useDPoP(context)` moved to the login builder |
| 7.5 | `DPoPException.UNSUPPORTED_ERROR` removed |
| 7.6 | `SSOCredentials.expiresIn: Int` → `expiresAt: Date` |
| 7.7 | `SecureCredentialsManager` `Auth0`-based constructors removed |

**Behavioral changes (surfaced, no code change):** default `minTtl` 0 → 60s, `clearCredentials()` now clears all storage, `CredentialsManager` uses the global executor.

### Structure

- `SKILL.md` — argument-based version resolution + validation, prerequisite gate, source fetch, grep-gated audit, build-until-green, migration summary
- `references/process.md` — version validation, prerequisite handling, build-system edge cases, MFA map, backend pattern, Gson notes, rollback
- `references/security.md` — Android security invariants
- `tests/` — eval suite (`evals.json`, `graders.json`, `graders.ts`, `benchmark-config.json`, `package.json`)

Registers the skill across the plugin manifests and READMEs.

### Testing

- `skillsaw --strict` passes: **A+, 0 errors, 0 warnings** (43 skills)
- All JSON validated; every v4 signature confirmed against the actual SDK source on `v4_development`

### Checklist

- [x] This change adds test coverage for new/changed functionality
- [x] Documentation added for new functionality
- [x] All active checks for tests, formatting, and security are passing

## Summary by CodeRabbit

* **New Features**
  * Added the `auth0-android-major-migration` skill to support upgrading Auth0.Android from v3 to v4.

* **Documentation**
  * Expanded the skill documentation with detailed upgrade steps, prerequisite checks, build-verification guidance, and a security checklist.
  * Updated Marketplace installation instructions to include the new skill install command.
  * Added supporting migration reference material for API/behavior changes and validation procedures.

### CODE REVIEW

**Comment by sanchitmehtagit** [comment]
## Tested against a sample scaffold

Ran the skill's full workflow against a realistic v3 scaffold — a Kotlin app on `com.auth0.android:auth0:3.19.0`, `minSdk 24` / Java 8, with `AuthManager.kt` deliberately mixing **changed** APIs (`SecureCredentialsManager(Auth0, …)`, `SSOCredentials.expiresIn`, `getCredentials` without `minTtl`, `clearCredentials`) and **unchanged** ones (no MFA / Management / Passkey / DPoP) to confirm the grep-gating only touches what's used.

**Workflow results:**

| Step | Result |
|---|---|
| Pre-flight | Clean-tree gate passed; backup branch created |
| Version resolution | Detected `3.19.0`; no arg → auto-resolved `4.0.0-beta.1`, flagged as pre-release |
| Prerequisite gate | Flagged all bumps; correctly tagged `minSdk 24 → 26` as a hard block (drops Android 7.1) |
| Source fetch | `4.0.0-beta.1` tag carries source; every signature confirmed |
| Audit | **Uses:** §7.6, §7.7, §8.1, §8.3 · **Skips:** §7.1–7.5, §8.4 |
| Dependency | `3.19.0 → 4.0.0-beta.1` (exact pin for pre-release) |
| Code changes | §7.7 constructor rebuilt to `AuthenticationAPIClient`; §7.6 `expiresIn: Int → expiresAt: Date` (+ `Date` import) |
| Behavioral | §8.1 (minTtl 0→60) and §8.3 (clearCredentials clears all) surfaced in summary |

**Correctness checks:** no removed-v3 API names remain in source; new `SecureCredentialsManager(apiClient, context, storage)` matches the v4 3-arg constructor read from source; all Auth0 symbols used in the file exist in the `4.0.0-beta.1` tree (no hallucinated APIs); `SharedPreferencesStorage` is built-in, so the §8.4 skip was correct.

Result: a minimal, correct changeset that touched only the four APIs the project actually uses.

File: plugins/auth0/skills/auth0-android-major-migration/references/process.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+The skill accepts an optional target version as an argument (`$ARGUMENTS`). Resolution happens in Step 2 of the main workflow.
+
+**When an argument is provided, it must pass all three checks before use:**
+
+1. **Exists** — the tag appears in the published release list:
+   ```bash
+   gh api repos/auth0/Auth0.Android/releases --paginate \
+     --jq '.[] | select(.draft==false) | .tag_name'
+   ```
```
[the remark this answers is not in the document]
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/references/process.md:18]: Fixed in 911dbb0 — added blank lines around the fences (including the adjacent groovy/kotlin pair at 96-97) and a `text` language specifier on the backend-pattern diagram.

File: plugins/auth0/skills/auth0-android-major-migration/references/security.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+### 1. No Secrets in Source Code
+
+```
+[ ] No client_secret in any Kotlin/Java/Gradle/XML file (native apps use PKCE, not secrets)
+[ ] No hardcoded tokens or API keys
+[ ] Auth0 domain and client ID live in strings.xml (public config, not secrets) — not hardcoded in source
+[ ] No Management API token embedded in the app (see §7.2 — Management API moves to a backend)
+[ ] No tokens logged to Logcat in production builds
+```
```
[the remark this answers is not in the document]
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/references/security.md]: Fixed in 911dbb0 — added `text` to the checklist fences and blank lines before the `bash` verification blocks.

File: plugins/auth0/skills/auth0-android-major-migration/references/security.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+- Verify credentials stored under v3 are still readable after the SDK update (storage format is backward compatible). If not, the app should prompt re-login rather than crash.
+- Verify biometric protection is preserved: the biometric `SecureCredentialsManager` constructor (with `FragmentActivity` + `LocalAuthenticationOptions`) must remain in use if it was before.
+- §8.1 (`minTtl` default 60s) and §8.3 (`clearCredentials()` clears all) are behavioral — confirm they don't break a security assumption the app relied on.
+
+### When Auth Flow APIs Change (§7.4 DPoP, scopes)
+
+- Verify the same scopes are still requested — especially `offline_access` for refresh tokens.
+- Verify the same audience is still passed for API access tokens.
+- Verify the callback URL scheme still matches the Auth0 Dashboard configuration.
+- §7.4: moving `useDPoP(context)` to the login builder must not silently drop DPoP — if the app used DPoP before, it must still configure it (now per-request).
```
[the remark this answers is not in the document]
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/references/security.md:79]: Skipping — the parallel "Verify …" phrasing is intentional in a security checklist where each item is an action to perform; the repetition aids scannability rather than hurting it. Pure prose-style nit, no functional impact.

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,549 @@
...
+```kotlin
+implementation("com.auth0.android:auth0:<TARGET_TAG>")
+```
+
+**Version catalog (`gradle/libs.versions.toml`):**
+```toml
+[versions]
+auth0 = "<TARGET_TAG>"
+```
+
```
[the remark this answers is not in the document]
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/SKILL.md:263]: Fixed in 911dbb0 — added blank lines between the bold intro text and the groovy/kotlin/toml fences.

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+    requires:
+      bins:
+        - gh
+---
+
+# Auth0.Android v4 Migration
+
+Migrates an existing Auth0.Android (`com.auth0.android:auth0`) v3 integration to v4. Every code change is gated on a search that confirms the project actually calls the affected API — if the project never uses `SecureCredentialsManager`, no `SecureCredentialsManager` code is touched. Changes follow the project's existing architecture (Kotlin or Java, callback or coroutine) and Android conventions.
+
+## Target version is argument-based
```
Comment (NandanPrabhu) [inline plugins/auth0/skills/auth0-android-major-migration/SKILL.md:21]: If target version is not provided in argument then we should prompt the user for the specific version. If user doesnt provide in the prompt version should be auto resolved to latest stable or beta

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+    .start(context, callback)
+```
+
+---
+
+#### 7.5 — `DPoPException.UNSUPPORTED_ERROR` removed
+
+**Applies if:** Step 5 found `DPoPException.UNSUPPORTED_ERROR` in the project's source files.
+
+With the minimum SDK raised to API 26, DPoP is supported on every API level v4 targets, so this constant was removed. Remove any reference to it — for example, a `when`/`if` branch or comparison that checked for `UNSUPPORTED_ERROR`. No replacement is needed; the unsupported-version case can no longer occur.
```
Comment (utkrishtsahu) [inline plugins/auth0/skills/auth0-android-major-migration/SKILL.md:357]: The reasoning here is factually off and could confuse anyone maintaining this later. In v3, DPoPException.UNSUPPORTED_ERROR carried the message "DPoP is not supported in versions below Android 9 (API level 28)." — but v4 only raises minSdk to 26, not 28. So "minSdk 26 means DPoP is supported on every API level v4 targets" isn't true; API 26 and 27 are still below 28. The constant was removed, yes, but the stated cause-and-effect is incorrect. Please either re-verify the actual reason DPoP no longer needs the unsupported branch in v4 (e.g. the implementation no longer gates on API 28 — I don't see any SDK_INT >= 28 check left in the v4 dpop/ source), and reword. The instruction (remove the reference) is fine; just the explanation is wrong.

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+    handle(error)
+}
+
+// v4 — the guard is no longer applicable; handle the remaining cases
+handle(error)
+```
+
+---
+
+#### 7.6 — `SSOCredentials.expiresIn` → `expiresAt` (`Int` → `Date`)
```
Comment (utkrishtsahu) [inline plugins/auth0/skills/auth0-android-major-migration/SKILL.md:373]: Confirmed correct — but one subtlety the migrator should know: in the v4 source the field is still @field:SerializedName("expires_in") even though the Kotlin property is expiresAt: Date. That's intentional in the SDK, but if anyone reading this skill greps the JSON wire format they may get confused. A one-line note ("the JSON key remains expires_in; only the Kotlin property and type changed") would pre-empt that.

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+| `unresolved reference: UNSUPPORTED_ERROR` | §7.5 — remove the reference |
+| `unresolved reference: expiresIn` on `SSOCredentials`, or `Int`/`Date` type mismatch | §7.6 — rename to `expiresAt` (now a `Date`) |
+| `none of the following functions can be called` / `too many arguments` on `SecureCredentialsManager(` | §7.7 — build `AuthenticationAPIClient(auth0)` first, pass it in |
+| `class … must override removeAll` (custom `Storage`) | §8.4 — the interface has a default impl; override only if needed |
+
+**Limit:** Up to **10 build-fix cycles**. If the build still fails after 10 attempts, stop and show the remaining errors with context — do not guess.
+
+---
+
+### Step 9 — Behavior & Default-Value Changes (Review, Usually No Code Change)
```
Comment (utkrishtsahu) [inline plugins/auth0/skills/auth0-android-major-migration/SKILL.md:453]: Step 9 is titled "Behavior & Default-Value Changes" but its subsections are numbered 8.1–8.4, and Step 10 then refers back to "§8.x". The audit table in Step 5 also points at "§8.1 / §8.3 / §8.4". It's internally consistent (everything calls them §8.x), so it works — but a reader will briefly trip on "Step 9 contains §8.x". Consider renumbering to §9.x, or relabel the step. Cosmetic only.

**Review by utkrishtsahu — CHANGES_REQUESTED** [review]

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md (hunk 5 above)
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/SKILL.md:21]: Yes if target version is not provided , We would auto resolve to the latest version without a prompt

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md (hunk 6 above)
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/SKILL.md:357]: https://github.com/auth0/Auth0.Android/blob/v4_development/V4_MIGRATION_GUIDE.md#dpopexceptionunsupported_error-removed

This was based on reference. Could you please recheck it? If not, let me know what the correct wording or reasoning should be.

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md (hunk 7 above)
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/SKILL.md:373]: Done in 38e3fc3 — added a note that the JSON wire key stays `expires_in` (`@field:SerializedName("expires_in")`); only the Kotlin property and type changed (`expiresIn: Int` → `expiresAt: Date`).

File: plugins/auth0/skills/auth0-android-major-migration/SKILL.md (hunk 8 above)
Comment (sanchitmehtagit) [reply plugins/auth0/skills/auth0-android-major-migration/SKILL.md:453]: Done in 38e3fc3 — renumbered the subsections §8.x → §9.x to match Step 9, and updated all cross-references (Step 5 audit table, Step 8 error table, Step 10 summary).

**Review by utkrishtsahu — APPROVED** [review]
